app/src/main/java/com/github/onlynotesswent/model/folder/FolderRepositoryFirestore.kt

```diff
@@ -74,80 +144,142 @@ class FolderRepositoryFirestore(private val db: FirebaseFirestore) : FolderRepos
+      val firestoreFolders =
+          withContext(Dispatchers.IO) {
+            db.collection(folderCollectionPath)
+                .get()
+                .await()
+                .documents
                 .mapNotNull { document -> documentSnapshotToFolder(document) }
                 .filter { it.userId == userId }
-        onSuccess(userFolders)
-      } else {
-        task.exception?.let { e: Exception ->
-          onFailure(e)
-          Log.e(TAG, "Failed to retrieve folders from user: ${e.message}")
-        }
-      }
+          }
+
+      // Sync Firestore with cache
+      val updatedFolders =
+          if (useCache) syncFoldersFirestoreWithCache(firestoreFolders, cachedFolders)
+          else firestoreFolders
+
+      onSuccess(updatedFolders)
+    } catch (e: Exception) {
+      Log.e(TAG, "Failed to retrieve folders from user: ${e.message}")
+      onFailure(e)
     }
   }
 
-  override fun getRootFoldersFromUid(
+  override suspend fun getRootFoldersFromUid(
       userId: String,
       onSuccess: (List<Folder>) -> Unit,
-      onFailure: (Exception) -> Unit
+      onFailure: (Exception) -> Unit,
+      useCache: Boolean
   ) {
-    db.collection(folderCollectionPath).get().addOnCompleteListener { task ->
-      if (task.isSuccessful) {
-        val rootFolders =
-            task.result.documents
+    try {
+      val cachedFolders: List<Folder> =
+          if (useCache) withContext(Dispatchers.IO) { folderDao.getRootFolders() } else emptyList()
+
+      // If device is offline, fetch from local database
+      if (!NetworkUtils.isInternetAvailable(context)) {
+        onSuccess(cachedFolders)
+        return
+      }
+
+      // If device is online, fetch from Firestore
+      val firestoreFolders =
+          withContext(Dispatchers.IO) {
+            db.collection(folderCollectionPath)
+                .get()
+                .await()
+                .documents
                 .mapNotNull { document -> documentSnapshotToFolder(document) }
                 .filter { it.userId == userId && it.parentFolderId == null } // Only root folders
-        onSuccess(rootFolders)
-      } else {
-        task.exception?.let { e: Exception ->
-          onFailure(e)
-          Log.e(TAG, "Failed to retrieve root folders from user: ${e.message}")
-        }
-      }
+          }
+
+      // Sync Firestore with cache
+      val updatedFolders =
+          if (useCache) syncFoldersFirestoreWithCache(firestoreFolders, cachedFolders)
+          else firestoreFolders
+
+      onSuccess(updatedFolders)
+    } catch (e: Exception) {
+      Log.e(TAG, "Failed to retrieve root folders from user: ${e.message}")
+      onFailure(e)
     }
   }
 
-  override fun updateFolder(folder: Folder, onSuccess: () -> Unit, onFailure: (Exception) -> Unit) {
+  override suspend fun updateFolder(
+      folder: Folder,
+      onSuccess: () -> Unit,
+      onFailure: (Exception) -> Unit,
+      useCache: Boolean
+  ) {
+    // Update the cache if needed
```

Comment:
If you use cache, and both caching and updating repository works, then you call onSuccess twice, which might cause unintended consequences (e.g. `onSuccess = navigationActions.goback()`)

Reply:
I will change this when I make the modifications for `useCache`. `addFolder`, `updateFolder` and anything that requires modifying a note will have to use cache by default. I will still keep the parameter in case we decide to change this in the future